Silence editor initialization act warnings triggered by asynchronous resolvers during unmount

dcalhoun opened the pull request:
## Description
Fixes #38332. Relates to https://github.com/WordPress/gutenberg/pull/38336. 

During editor initialization,[ asynchronous store resolvers](https://github.com/WordPress/gutenberg/blob/57da3c91a166d917a2a9de98177be9c3dfe07ee5/packages/editor/src/components/provider/index.native.js#L370) leverage `apiFetch` to fetch data from the REST API. The resolvers also [rely upon `setTimeout`](https://github.com/WordPress/gutenberg/blob/e1978ea1bb20e99be9f64a189ac5f898a7cee594/packages/data/src/redux-store/index.js#L414-L428) to run at the end of the current JavaScript block execution. In order to prevent `act` warnings triggered by updates to the React tree, we manually tick fake timers and flush micro tasks before proceeding. 

My theory as to why the specific test case in #38332 exposed further `act` warnings is because it is a simplified test case that has no asynchronous logic of its own, aside from editor initialization. This is different than most of our existing tests. Our existing tests await the resolution of their own asynchronous logic, which allowed an opportunity for the editor initialization asynchronous logic to resolve undetected. 

https://github.com/WordPress/gutenberg/blob/74722f69be32bad1b55177312d1865bf5589b7fb/packages/editor/src/components/provider/index.js#L91

The `act` warning appears to originate from the above line, which triggers during unmount. Components are unmounted after each test, which would explain why the `act` warning is triggered. Flushing of micro tasks is somewhat of a "sledgehammer" to address the `act` warnings, rather than a targeted awaiting of a specific fetch request. However, it has been difficult to identify the request to await and the request resolution appears to have no meaningful impact on the subject of our tests using `initializeEditor`, which do not focus on testing editor dismount. 

## Testing Instructions
1. Verify that the `act` warnings are not displayed following the reproduction instructions from https://github.com/WordPress/gutenberg/issues/38332.
2. Run `npm run native test` and observe that all tests pass.

## Screenshots <!-- if applicable -->
n/a

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [x] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

dcalhoun committed:
Fix editor initialization act warnings

During editor initialization, asynchronous store resolvers leverage
`apiFetch` to fetch data from the REST API. The resolvers also rely upon
`setTimeout` to run at the end of the current JavaScript block
execution. In order to prevent "act" warnings triggered by updates to
the React tree, we manually tick fake timers and await the resolution of
`apiFetch` before proceeding.

fluiddot commented on test/native/helpers.js:
```diff
@@ -43,19 +43,28 @@ provideToNativeHtml.mockImplementation( ( html ) => {
+	// before proceeding.
 	jest.useFakeTimers( 'legacy' );
+	const mockApiResponse = Promise.resolve();
+	jest.doMock( '@wordpress/data-controls', () => {
+		const dataControls = jest.requireActual( '@wordpress/data-controls' );
+		return {
+			...dataControls,
+			apiFetch: jest.fn( () => mockApiResponse ),
+		};
+	} );
```
I tried to debug `apiFetch` from `@wordpress/data-controls` but I couldn't manage to identify any calls during initialization. In fact, if you comment this code block the tests pass too 🤔 , which makes me wonder if we need to mock it, wdyt?

On the other hand, I understand that API requests are already being omitted as they are mocked on the Jest setup ([reference](https://github.com/WordPress/gutenberg/blob/fa5c739fc4f38aa92587a2cb31b4b0998bb6bf81/test/native/setup.js#L45-L50)), so maybe we don't need to mock them for this purpose.

dcalhoun replied:
> I tried to debug `apiFetch` from `@wordpress/data-controls` but I couldn't manage to identify any calls during initialization. In fact, if you comment this code block the tests pass too 🤔 , which makes me wonder if we need to mock it, wdyt?

You are absolutely correct. It would appear the successful run was merely a side effect of the presence of `await act (() => anyPromiseResolve);`.  The specific mock was irrelevant. Good catch. 

> On the other hand, I understand that API requests are already being omitted as they are mocked on the Jest setup ([reference](https://github.com/WordPress/gutenberg/blob/fa5c739fc4f38aa92587a2cb31b4b0998bb6bf81/test/native/setup.js#L45-L50)), so maybe we don't need to mock them for this purpose.

Agreed, this mock merely hoisted the mocking further up from `@wordpress/api-fetch` to `@wordpress/data-controls`. Your first point regarding commenting the mock out makes this subject moot, however.

fluiddot reviewed (commented):
> My theory as to why the specific test case in #38332 exposed further `act` warnings is because it is a simplified test case that has no asynchronous logic of its own, aside from editor initialization. This is different than most of our existing tests. Our existing tests await the resolution of their own asynchronous logic, which allowed an opportunity for the editor initialization asynchronous logic to resolve undetected.

I agree with your theory, I don't think we would bump into the `act` warnings in asynchronous tests, as the potential React tree updates produced by the selectors, related to the `act` warnings, would be absorbed under `waitFor` calls.

> That said, it is possible the changes in this PR awaiting the resolution of `apiFetch` is, again, another "red herring" and further masking of the a deeper issue. It is difficult for me to determine for certain due to the complexity of the store resolvers and the large React tree mounted by `initializeEditor`. However, given store resolvers _do_ leverage `apiFetch` and _do_ leverage `setTimeout`, it makes sense to me that these changes address the `act` warnings.

I think this is a very good approach, as we'd cover all asynchronous logic related to selector resolvers. I also checked the requests that are made upon initialization and most of them are not supported by the native version ([reference](https://github.com/WordPress/gutenberg/blob/fa5c739fc4f38aa92587a2cb31b4b0998bb6bf81/packages/react-native-editor/src/api-fetch-setup.js#L7-L12)). So in theory, we shouldn't expect any impact on the tests by mocking the requests, except for tests that rely on `media` endpoint, although in this case most likely the tests would provide its own mocked response.

**Requests with unsupported endpoints:**
- `/wp/v2/settings?_locale=user`
- `/?_locale=user`
- `/wp/v2/pages?_locale=user`
- `/wp/v2/posts/post-id-<UNIQUE_UUID>?_locale=user`

**Requests with supported endpoints:**
- `/wp/v2/media?_locale=user`

This also made me think if we should maybe rethink how we initialize the editor for the tests, as we're omitting the setup code covered in [`packages/react-native-editor/src/setup.js`](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/src/setup.js). I'm wondering if we should use the entry point [`registerGutenberg` function](https://github.com/WordPress/gutenberg/blob/fa5c739fc4f38aa92587a2cb31b4b0998bb6bf81/packages/react-native-editor/src/index.js#L31), for initializing the editor instead of rendering the Editor component, wdyt?

fluiddot commented on test/native/helpers.js:
```diff
@@ -81,6 +90,9 @@ export function initializeEditor( props, { component = Editor } = {} ) {
 	// Advance all timers allowing store resolvers to resolve.
 	act( () => jest.runAllTimers() );
 
+	// Await resolution of API fetches within store resolvers.
+	await act( () => mockApiResponse );
+
```
Another option for waiting for the store resolvers would be using `setImmediate`, similarly as the RN testing library waits for the microtasks when the test finishes ([reference](https://github.com/callstack/react-native-testing-library/blob/a010ffdbca906615279ecc3abee423525e528101/src/flushMicroTasks.js#L15-L23)).

```js
	// Await microtasks related to resolution of store resolvers.
	await act( () => new Promise( ( resolve ) => setImmediate( resolve ) ) );

```

dcalhoun replied:
I like this approach with a strategically placed `setImmediate` better than https://github.com/WordPress/gutenberg/pull/38336, where we silence all warnings. It is still in some ways masking a deeper origin, but I feel it is not as far reaching, which I consider a good approach. 👍🏻

dcalhoun committed:
Resolve API fetches during editor initialization

dcalhoun commented:
> I think this is a very good approach, as we'd cover all asynchronous logic related to selector resolvers. I also checked the requests that are made upon initialization and most of them are not supported by the native version ([reference](https://github.com/WordPress/gutenberg/blob/fa5c739fc4f38aa92587a2cb31b4b0998bb6bf81/packages/react-native-editor/src/api-fetch-setup.js#L7-L12)). 

I had not come across this aspect of the API fetch setup. Thank you for sharing this; it provides a lot of great context. 

> So in theory, we shouldn't expect any impact on the tests by mocking the requests, except for tests that rely on `media` endpoint, although in this case most likely the tests would provide its own mocked response.

As your other point below references — regarding our approach to initializing the editor — the logic "disabling" the unsupported API endpoints is not run in our current test environment set up. So, mocks or lack thereof _would_ impact the tests currently. 

I [refactored this PR's original fetch mocks](https://github.com/WordPress/gutenberg/commit/b2880e18861daa2438dfe58fd091d257db4ff69f) to target the correct fetch logic, and it does, in fact, have an impact on the outcome. If one checks out that branch/commit, the example tests from https://github.com/WordPress/gutenberg/issues/38332 pass. If one removes the fetch mock resolution, the tests continue passing — presumably because the mocked fetch merely never resolves. If one resolves the fetch without `act`, the tests fail — presumably because the resolved fetches cause React tree updates. 

<details><summary>Remove Fetch Mock Resolution</summary>

```diff
diff --git a/test/native/helpers.js b/test/native/helpers.js
index c9d7198d4c..655eff718e 100644
--- a/test/native/helpers.js
+++ b/test/native/helpers.js
@@ -90,11 +90,11 @@ export async function initializeEditor( props, { component = Editor } = {} ) {
 	act( () => jest.runAllTimers() );
 
 	// Await resolution of API fetches within store resolvers.
-	await act( async () => {
-		apiFetchResolves.forEach( ( resolve ) => {
-			resolve();
-		} );
-	} );
+	// await act( async () => {
+	// 	apiFetchResolves.forEach( ( resolve ) => {
+	// 		resolve();
+	// 	} );
+	// } );
 
 	// Restore the default timer APIs for remainder of test arrangement, act, and
 	// assertion.
```

</details>

<details><summary>Fetch Resolution w/o act</summary>

```diff
diff --git a/test/native/helpers.js b/test/native/helpers.js
index c9d7198d4c..6001ed2003 100644
--- a/test/native/helpers.js
+++ b/test/native/helpers.js
@@ -90,10 +90,8 @@ export async function initializeEditor( props, { component = Editor } = {} ) {
 	act( () => jest.runAllTimers() );
 
 	// Await resolution of API fetches within store resolvers.
-	await act( async () => {
-		apiFetchResolves.forEach( ( resolve ) => {
-			resolve();
-		} );
+	apiFetchResolves.forEach( ( resolve ) => {
+		resolve();
 	} );
 
 	// Restore the default timer APIs for remainder of test arrangement, act, and
```

</details>

> This also made me think if we should maybe rethink how we initialize the editor for the tests, as we're omitting the setup code covered in [`packages/react-native-editor/src/setup.js`](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/src/setup.js). I'm wondering if we should use the entry point [`registerGutenberg` function](https://github.com/WordPress/gutenberg/blob/fa5c739fc4f38aa92587a2cb31b4b0998bb6bf81/packages/react-native-editor/src/index.js#L31), for initializing the editor instead of rendering the Editor component, wdyt?

That is an interesting thought. First, I wonder how we might use `registerGutenberg` as the test environment entry point alongside of Testing Library's `render`. The former ultimately registers a component with `AppRegistry` to be invoked from the native side. I suppose we could mock `AppRegistry` in some way? 

Second, I wonder if `registerGutenberg`'s logic is too complex, large, or unrelated to the subject of mosts of these tests, e.g. individual edit views for blocks. I wonder if we'd be better off unit testing `registerGutenberg` separately. 

Ultimately, I think it would be most beneficial to implement a solution that at least somewhat targets the apparent origin of the `act` warnings — asynchronous resolvers during editor initialization. I think that may likely be either https://github.com/WordPress/gutenberg/pull/38344#discussion_r795786205 or awaiting mocked fetches in some manner. 

While I would love to identify the perfect approach to awaiting the specific fetches, I do not want invest too much time chasing perfection. Do you think https://github.com/WordPress/gutenberg/pull/38344#discussion_r795786205 is good enough?

fluiddot commented:
> I refactored this PR's original fetch mocks to target the correct fetch logic, and it does, in fact, have an impact on the outcome. If one checks out that branch/commit, the example tests from #38332 pass. If one removes the fetch mock resolution, the tests continue passing — presumably because the mocked fetch merely never resolves. If one resolves the fetch without act, the tests fail — presumably because the resolved fetches cause React tree updates.

That's interesting, I'm wondering what updates exactly are being done by the selectors that make API requests 🤔 .

While testing the changes from https://github.com/WordPress/gutenberg/commit/b2880e18861daa2438dfe58fd091d257db4ff69f, I realized that the following approach works too:
```diff
diff --git a/test/native/helpers.js b/test/native/helpers.js
index c9d7198d4c151638609739bd98a1c673731944a4..8d0ab8d0227a401f7e4c939bcf151c7ec6b551ee 100644
--- a/test/native/helpers.js
+++ b/test/native/helpers.js
@@ -58,12 +58,7 @@ export async function initializeEditor( props, { component = Editor } = {} ) {
 	// we manually tick fake timers and await the resolution of `apiFetch`
 	// before proceeding.
 	jest.useFakeTimers( 'legacy' );
-	const apiFetchResolves = [];
-	apiFetch.mockImplementation( () => {
-		return new Promise( ( resolve ) => {
-			apiFetchResolves.push( resolve );
-		} );
-	} );
+	apiFetch.mockImplementation( () => Promise.resolve() );
 
 	// Arrange
 	const EditorComponent = component;
@@ -90,11 +85,7 @@ export async function initializeEditor( props, { component = Editor } = {} ) {
 	act( () => jest.runAllTimers() );
 
 	// Await resolution of API fetches within store resolvers.
-	await act( async () => {
-		apiFetchResolves.forEach( ( resolve ) => {
-			resolve();
-		} );
-	} );
+	await act( async () => {} );
 
 	// Restore the default timer APIs for remainder of test arrangement, act, and
 	// assertion.
```

Which makes me think that the line `await act( async () => {} );` might be performing something similar to `await act( () => new Promise( ( resolve ) => setImmediate( resolve ) ) );` and acknowledging the React tree updates.

> That is an interesting thought. First, I wonder how we might use registerGutenberg as the test environment entry point alongside of Testing Library's render. The former ultimately registers a component with AppRegistry to be invoked from the native side. I suppose we could mock AppRegistry in some way?

Yeah, in fact, we're already doing it in [`packages/react-native-editor/src/test/index.test.js`](https://github.com/WordPress/gutenberg/blob/65f7c31da4b5e0e89ab8b3b6997361834ddd0b00/packages/react-native-editor/src/test/index.test.js#L19-L31).

> Second, I wonder if registerGutenberg's logic is too complex, large, or unrelated to the subject of mosts of these tests, e.g. individual edit views for blocks. I wonder if we'd be better off unit testing registerGutenberg separately.

Right, probably it's too much for testing the editor, although, on the other hand, it would be the closest logic to how the app initializes the editor. In any case, I wouldn't focus on this for the purpose of this PR, just wanted to raise it in case we could get ideas/clues for solving this issue.

> Ultimately, I think it would be most beneficial to implement a solution that at least somewhat targets the apparent origin of the act warnings — asynchronous resolvers during editor initialization. I think that may likely be either #38344 (comment) or awaiting mocked fetches in some manner.

I totally agree, both options are similar so I don't have a strong preference. Regarding waiting for mocked fetches, from the following code:
```js
	await act( async () => {
		apiFetchResolves.forEach( ( resolve ) => {
			resolve();
		} );
	} );
```

I'm wondering if this logic is actually waiting for the fetches, as the code within `act` function is not asynchronous, right? I might be overlooking anything so please let me know if my statement is not accurate 😅 . In any case, your approach from https://github.com/WordPress/gutenberg/commit/b2880e18861daa2438dfe58fd091d257db4ff69f is a good workaround.

> While I would love to identify the perfect approach to awaiting the specific fetches, I do not want invest too much time chasing perfection. Do you think #38344 (comment) is good enough?

Yep, I think so, at least if we guarantee that the API requests are solved instantly. By advancing the timers and mocking requests, it should be safe to use this approach, as we're enforcing the selector resolvers to be executed in two cycles:

https://github.com/WordPress/gutenberg/blob/65f7c31da4b5e0e89ab8b3b6997361834ddd0b00/packages/data/src/redux-store/index.js#L414-L428

- **First cycle:** `await fulfillResolver` is called and triggers the API requests, however as it's called asynchronously it most likely requires an extra cycle to finish. After checking the [Promises/A+ specification](https://promisesaplus.com/#point-67) for gathering some insights about how the `await` might be running the asynchronous code, I have the impression that its implementation is probably using either macrotasks or microtasks, which would prove that it requires another cycle.
- **Second cycle:** API requests are resolved instantly so in this cycle all logic related to the resolver should have been executed.

fluiddot commented:
> That's interesting, I'm wondering what updates exactly are being done by the selectors that make API requests 🤔 .

I did a quick check and spotted that the following line is the actual reason for having to wait for a cycle:
https://github.com/WordPress/gutenberg/blob/74722f69be32bad1b55177312d1865bf5589b7fb/packages/editor/src/components/provider/index.js#L91

If you comment that line, the tests pass even with the workaround we have in `trunk` ([reference](https://github.com/WordPress/gutenberg/blob/trunk/test/native/helpers.js)). However, I haven't found yet how by setting [the `isReady` property](https://github.com/WordPress/gutenberg/blob/74722f69be32bad1b55177312d1865bf5589b7fb/packages/editor/src/store/reducer.js#L242-L262) to `false` when dispatching `TEAR_DOWN_EDITOR` action, produces the React tree updates 🤔 .

dcalhoun commented:
> That's interesting, I'm wondering what updates exactly are being done by the selectors that make API requests 🤔 .

**test # 1:**
- `/wp/v2/settings`
- `/`
- `/wp/v2/media`
- `/wp/v2/pages`
- `/wp/v2/posts/post-id-4728b0f1-358f-425e-9f25-b1583870c68a`

**test # 2:**
- `/wp/v2/posts/post-id-cae60b37-5965-4cfc-9357-63246edcaddf`

These are the endpoints fetched during the sample tests. As you mentioned earlier, only the `/wp/v2/media` endpoint is allowed by the native editor. Since we are not running the setup code disallowing them, they arrive to `apiFetch`. If we _did run the setup code_, it would [trigger a `Promise.reject`](https://github.com/WordPress/gutenberg/blob/74722f69be32bad1b55177312d1865bf5589b7fb/packages/react-native-editor/src/api-fetch-setup.js#L22), which we would still need to await to avoid `act` warnings regardless. Presumably, the second test invokes fewer fetches due to cached resolvers from the first test? 

> Which makes me think that the line `await act( async () => {} );` might be performing something similar to `await act( () => new Promise( ( resolve ) => setImmediate( resolve ) ) );` and acknowledging the React tree updates.

I agree. It does appear that these produce the same results.  

> I'm wondering if this logic is actually waiting for the fetches, as the code within `act` function is not asynchronous, right? I might be overlooking anything so please let me know if my statement is not accurate 😅 . In any case, your approach from [b2880e1](https://github.com/WordPress/gutenberg/commit/b2880e18861daa2438dfe58fd091d257db4ff69f) is a good workaround.

I suppose it depends upon what you mean by "waiting for the fetches." I do not believe it waits on them in the sense that there is a return value from within `async () => {...}`. However, I believe it is awaiting `async () => {...}` itself. Removing `await` and `async` from this code block reintroduces the `act` warnings. 

The approach here was taken from https://github.com/testing-library/react-testing-library/issues/542#issuecomment-562968846 and [Sunil's documentation](https://github.com/threepointone/react-act-examples/blob/06d382801a9de840cbc6ec18023206705106a354/sync.md). I believe Sunil was heavily involved in the design of `act` and thus I trust those resources. 

> If you comment that line, the tests pass even with the workaround we have in `trunk` ([reference](https://github.com/WordPress/gutenberg/blob/trunk/test/native/helpers.js)). However, I haven't found yet how by setting [the `isReady` property](https://github.com/WordPress/gutenberg/blob/74722f69be32bad1b55177312d1865bf5589b7fb/packages/editor/src/store/reducer.js#L242-L262) to `false` when dispatching `TEAR_DOWN_EDITOR` action, produces the React tree updates 🤔 .

Yep, I arrived at the same placed. I have not determined how `isReady` is triggering these updates either. 😕

fluiddot commented:
> I suppose it depends upon what you mean by "waiting for the fetches." I do not believe it waits on them in the sense that there is a return value from within `async () => {...}`. However, I believe it is awaiting `async () => {...}` itself. Removing `await` and `async` from this code block reintroduces the `act` warnings.
> 
> The approach here was taken from [testing-library/react-testing-library#542 (comment)](https://github.com/testing-library/react-testing-library/issues/542#issuecomment-562968846) and [Sunil's documentation](https://github.com/threepointone/react-act-examples/blob/06d382801a9de840cbc6ec18023206705106a354/sync.md). I believe Sunil was heavily involved in the design of `act` and thus I trust those resources.

Ah, I've read through the documentation you shared, and now I understand a bunch of things about how `act` works that didn't know, thanks for sharing it 🙇 .

_From [Sunil's documentation](https://github.com/threepointone/react-act-examples/blob/06d382801a9de840cbc6ec18023206705106a354/sync.md)_:
> This time, the test passes, and the warning's disappeared. Brilliant. Of note, even thought it might appear like resolve(42) is synchronous, we use the async version to make sure microtasks are flushed before releasing scope, preventing the warning. Neat.

This clearly explains how this approach helps out for addressing the issue. However, in relation to the use of `__experimentalTearDownEditor`, I'm wondering if the wait for microtasks we're doing is actually not related to fetching Promises but the tear-down updates 🤔 .

> Yep, I arrived at the same placed. I have not determined how isReady is triggering these updates either. 😕

A mystery indeed 😄 .

dcalhoun committed:
Flush micro tasks during editor initialization

The store resolvers perform several API fetches during editor
initialization. The most straightforward approach to ensure all of them
resolve before we consider the editor initialized is to flush micro
tasks, similar to the approach found in `@testing-library/react-native`.
https://github.com/callstack/react-native-testing-library/blob/a010ffdbca906615279ecc3abee423525e528101/src/flushMicroTasks.js#L15-L23

dcalhoun commented:
@fluiddot I simplified the approach for silencing the `act` warnings triggered during component unmount based upon our most recent discussion. Please let me know if you believe it is appropriate and suitable for now.

fluiddot reviewed (approved):
> @fluiddot I simplified the approach for silencing the act warnings triggered during component unmount based upon our most recent discussion. Please let me know if you believe it is appropriate and suitable for now.

As you outlined in the PR's description:

> However, it has been difficult to identify the request to await and the request resolution appears to have no meaningful impact on the subject of our tests using initializeEditor, which are do not test unmounting the editor.

I agree that, although we're not actually waiting for the request resolution, I don't foresee any major impact on the test results with this approach by silencing the `act` warnings. Therefore, this workaround looks great to me 🎊 !

Thank you very much @dcalhoun for the hard work on investigating and providing helpful solutions for the issue, I really appreciate your help 🙇 .

fluiddot commented:
@dcalhoun it would be great if we open a PR in GB-mobile to let this fix be used in integration tests in that repo, wdyt?
